### PR TITLE
Java racing car 정현모

### PR DESCRIPTION
**Pull request by gusah009**
후진 없는 레이싱 게임
약 2시간 소요

### CODE REVIEW

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+      printGuideMessage("예시) jhm,cwc,ljh,shk");
+      String playerInput = getPlayerInput();
+      setPlayerCars(playerInput);
+    } catch (Exception e) {
+      e.printStackTrace();
+      printErrorMessage();
+      return FAIL;
+    }
+    return SUCCEED;
+  }
```
Comment (woo-chang) [inline gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:47]: 입력 받는 것에 있어서 요구사항과 조금 다르게 동작하는 것 같습니다!
Screen Shot 2022-07-24 at 5 30 05 PM
예외 메시지를 안보이도록 하는 것과 비정상적인 입력에 대한 처리가 추가되어야 할 것 같아용
Screen Shot 2022-07-24 at 5 29 54 PM

File: gusah009/java-racing-car/src/main/java/javablooming/Main.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import javablooming.service.GamePlay;
+
+public class Main {
+
+  public static void main(String[] args) {
+    GamePlay game = new GamePlay();
+    game.input();
+    game.play();
+    game.printResult();
+  }
```
Comment (woo-chang) [inline gusah009/java-racing-car/src/main/java/javablooming/Main.java:12]: Main문만 보고도 어떤 동작을 수행하는지 짐작할 수 있습니다. 좋아요👍

File: gusah009/java-racing-car/src/main/java/javablooming/domain/Car.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  String name;
+  int position = 0;
+
+  public Car(String name) {
+    if (name.length() > 5) {
+      throw new IllegalArgumentException();
+    }
+    this.name = name;
+  }
+
```
Comment (woo-chang) [inline gusah009/java-racing-car/src/main/java/javablooming/domain/Car.java:14]: 생성자에서 확인하기보다는 생성자 호출 전 검증을 진행하고 올바른 이름만 들어올 수 있도록 하는 방법은 어떨까요?

File: gusah009/java-racing-car/src/main/java/javablooming/domain/Car.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    return name;
+  }
+
+  public int getPosition() {
+    return position;
+  }
+
+  public void moveForward() {
+    position++;
+  }
```
Comment (woo-chang) [inline gusah009/java-racing-car/src/main/java/javablooming/domain/Car.java:25]: `setter`로 변경할 수 있는데 의도를 드러내는 메서드를 생성하는 것은 좋은 방법인것 같습니다!

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  int getMoveCount() {
+    return moveCount;
+  }
+
+  public void input() {
+    while (tryInputCarNames() == FAIL)
+      ;
+    while (tryInputMoveCount() == FAIL)
+      ;
+  }
```
Comment (woo-chang) [inline gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:35]: 반복적인 입력 처리에 대한 고민이 많았는데 되게 직관적인 것 같습니다. 참고하겠습니다👍

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+          String.format("%5s", playerCar.getName()) + " : " + getPositionToDash(playerCar));
+    }
+    System.out.println();
+  }
+
+  private String getPositionToDash(Car playerCar) {
+    // position의 크기만큼 대쉬("-") 생성
+    return new String(new char[playerCar.getPosition()])
+        .replace("\0", "-");
+  }
```
Comment (woo-chang) [inline gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:110]: java 11부터 지원하는 것 같지만 `"-".repeat(횟수)` 이런 방법도 존재하는 걸 확인했습니다. 좋은 정보인 것 같아서 공유합니다!

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+  private void playEachRacing() {
+    for (Car playerCar : playerCars) {
+      if (getEachCarRacingResult() == FORWARD) {
+        playerCar.moveForward();
```
Comment (woo-chang) [inline gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:115]: 메서드로 `부울 값`을 반환하기보다는 `열거 타입`의 비교가 들어가면 가독성이 좋아지는 것 같아요. 배워갑니다👍

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        firstCarList.add(playerCar);
+      } else if (firstCarList.get(0).getPosition() < playerCar.getPosition()) {
+        firstCarList.clear();
+        firstCarList.add(playerCar);
+      } else if (firstCarList.get(0).getPosition() == playerCar.getPosition()) {
+        firstCarList.add(playerCar);
+      }
+    }
+    return firstCarList;
+  }
```
Comment (woo-chang) [inline gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:156]: 요구조건에서 `depth`는 `2`까지 허용하고 `else` 사용을 금지한다고 봤었는데 요구조건과 조금 다른 부분이 존재하는 것 같습니다.

File: gusah009/java-racing-car/src/main/java/javablooming/util/GameUtil.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    }
+  }
+
+  public enum RacingResult {
+    FORWARD(true), STAY(false);
+
+    RacingResult(boolean b) {
+    }
+  }
+}
```
Comment (woo-chang) [inline gusah009/java-racing-car/src/main/java/javablooming/util/GameUtil.java:22]: 적절한 열거 타입의 활용 매우 좋습니다. 저도 적극적으로 활용하도록 해보겠습니다.

File: gusah009/java-racing-car/src/main/java/javablooming/domain/Car.java (hunk 3 above)
Comment (gusah009) [reply gusah009/java-racing-car/src/main/java/javablooming/domain/Car.java:14]: 저는 **입력이나 유효성 책임을 가지는 클래스를 따로 만들지 않았기 때문에** `Car` 클래스가 `carName`에 대한 모든 책임을 가지고 있다고 생각했습니다. 실제로 [이 토론](https://stackoverflow.com/questions/49335072/how-to-validate-constructor-parameters-on-java)에서도 문제가 없다고 하고 있어서 위 방법도 괜찮다고 생각합니다!

다만 입력이나 유효성 책임을 가지는 클래스를 생성한다면 위와 같은 작업은 해당 클래스에 책임을 위임하는 것도 괜찮을 것 같습니다. 좋은 의견 감사합니다! ☺️

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java (hunk 1 above)
Comment (gusah009) [reply gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:47]: 확인했습니다! 이름이 5자 이하이기만 하면 되는 줄 알았는데, 0자인 경우가 있었네요.

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java (hunk 6 above)
Comment (gusah009) [reply gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:110]: 헐 이걸로 바꿔야겠습니다! 감사합니다 😆

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java (hunk 8 above)
Comment (gusah009) [reply gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:156]: image

저는 [이 글](https://zzang9ha.tistory.com/307)을 보고 depth가 2라는 조건은 충족한다고 생각했습니다. 

`else`사용을 금지한다는 것이 `else` 에서 나머지 모든 조건을 수용하는 코드를 막기 위해서 금지하는 것으로 알고 있어서 `else if`는 괜찮다고 생각했는데, 찾아보니 `else if` 역시 코드의 가독성을 해쳐서 쓰지 않는 것이 좋은 것 같습니다. 수정하겠습니다. 감사합니다!

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+      printErrorMessage();
+      return FAIL;
+    }
+    return SUCCEED;
+  }
+
+  private void setMoveCount(String playerInput) {
+    moveCount = Integer.parseInt(playerInput);
+  }
+
```
Comment (JeonghakLee) [inline gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:86]: 입력이 음의 정수일 때의 예외처리도 있으면 좋을 것 같습니다!

**Review by woo-chang — APPROVED** [review]

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+    return castCarsName(firstCarNameList);
+  }
+
+  List<Car> getFirstCarList() {
+    int maxPosition = getMaxPosition(playerCars);
+    return playerCars.stream()
+        .filter(playerCar -> playerCar.getPosition() == maxPosition)
+        .toList();
+  }
+
```
Comment (JeonghakLee) [inline gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:149]: 이 부분에선 람다를 활용한 것이 확실히 코드가 간결하고 가독성이 좋아 보이네요

File: gusah009/java-racing-car/src/test/java/javablooming/domain/CarTest.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    //given
+    String carName = "123456789";
+
+    //when
+    boolean carNameValid = isCarNameValid(carName);
+
+    //then
+    assertThat(carNameValid).isEqualTo(false);
+  }
+
```
Comment (JeonghakLee) [inline gusah009/java-racing-car/src/test/java/javablooming/domain/CarTest.java:129]: 테스트를 할 때 해당 메서드는 경계 값 내에 포함되는지를  검증을 하고 있어 경계에 있는 값들과 가운데 정도만 테스트 해도 충분하다는 생각이 듭니다. 만약 length가 100까지 가능하다고 하면 100개의 케이스를 다 작성하는 것은 오버 엔지니어링이 아닌가 하는 생각이 들었습니다. 이에 관해서는 어떻게 생각하시나요?

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java (hunk 10 above)
Comment (gusah009) [reply gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:86]: 음의 정수일 때의 예외처리가 없었네요! 감사합니다 (_ _)

File: gusah009/java-racing-car/src/test/java/javablooming/domain/CarTest.java (hunk 12 above)
Comment (gusah009) [reply gusah009/java-racing-car/src/test/java/javablooming/domain/CarTest.java:129]: 동의합니다. 전부 테스트 하는건 오버엔지니어링 같습니다. 

유효한 경우가 5개 밖에 없어서 이렇게 하긴 했지만 말씀하신대로 100개 1000개를 넘어가면 `edge case`와 중간 값 몇 개만 추려서 테스트 하는 게 좋아보입니다!

File: gusah009/java-racing-car/src/test/java/javablooming/service/GamePlayTest.java [hunk 13]
```diff
@@ -190,6 +205,8 @@ void getRacingResultWhen9() {
   }
 
   private void setSystemInput(String data) {
-    System.setIn(new ByteArrayInputStream(data.getBytes()));
+    ByteArrayInputStream userInputStream = new ByteArrayInputStream(data.getBytes());
+    SCANNER = new Scanner(userInputStream);
+    System.setIn(userInputStream);
   }
```
Comment (woo-chang) [inline gusah009/java-racing-car/src/test/java/javablooming/service/GamePlayTest.java:211]: 이런 방법으로 입력에 대한 테스트가 가능하군요 !

**Review by woo-chang — APPROVED** [review]
😎👍👍

File: gusah009/java-racing-car/src/test/java/javablooming/service/GamePlayTest.java (hunk 13 above)
Comment (gusah009) [reply gusah009/java-racing-car/src/test/java/javablooming/service/GamePlayTest.java:211]: 넵 저 부분을 하면서 생긴 시행착오를 블로그에 써놨습니다! 심심할 때 읽어봐여~ 

[System.SetIn과 Scanner를 이용한 Junit 다중 테스트 시 NoSuchElementException](https://gusah009.github.io/etc/System_SetIn-Junit-Scanner/)

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java (hunk 10 above)
Comment (JeonghakLee) [reply gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:86]: 오 이런 메서드도 지원하네요

**Review by JeonghakLee — APPROVED** [review]

File: gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java (hunk 5 above)
Comment (shkisme) [reply gusah009/java-racing-car/src/main/java/javablooming/service/GamePlay.java:35]: 오 이렇게 깔끔한 방법이... 참고하겠습니다!

File: gusah009/java-racing-car/src/test/java/javablooming/service/GamePlayTest.java (hunk 13 above)
Comment (shkisme) [reply gusah009/java-racing-car/src/test/java/javablooming/service/GamePlayTest.java:211]: 헐 이런 방법이...! 👍👍

**Review by shkisme — APPROVED** [review]
많이 배워갑니다..👍
